### ISSUE TITLE
Make sure input files are in the correct format

### ISSUE THREAD

**Issue by janstrauss1** [ridgelab/JustOrthologs#8]
Hi there,

I've downloaded genome fasta and gff files from NCBI using [ncbi-genome-download](https://github.com/kblin/ncbi-genome-download) scripts and I'd now like to use `JustOrthologs` to detect orthologs for multiple species.

Unfortunately, I'm receiving the error `Make sure input files are in the correct format`.

It is even thrown when running the example codes

`python wrapper.py -g1 smallTest/wrapperTest/small_pan.gff3 -g2 smallTest/wrapperTest/small_human.gff3 -r1 smallTest/wrapperTest/small_pan.fasta.gz -r2 smallTest/wrapperTest/small_human.fasta.gz -all -o output`

and

`bash run_multiple_species.sh -i smallTest/wrapperTest/ -o outputDir/ -e output_ortholog_groups -f .fasta.gz -g .gff3`

Any ideas what the problem might be?

Many thanks in advance for your help!

Jan

**Comment by ridgelab** [comment]
Hello Jan,

Thank you for reaching out. I just cloned the Git repo onto a different machine to make sure that everything was still working. I copied the command that you wrote above for the wrapper (that you copied from the README), and I was unable to replicate the error. I do know that some people have had issues when the permissions aren't set for writing in the directory because the wrapper attempts to create a couple temporary files that are later deleted. If those files are empty, then that error message might appear (because the temp files should be empty only if the input files are incorrectly formatted or the files could not be created, likely due to permissions). Other than that, I'm not sure what would be causing this issue.

**Comment by janstrauss1** [comment]
Hi @ridgelab,

thanks a lot for your feedback!
Permissions are all set for writing in the directory as far as I can see.

Also, `python justOrthologs.py -q smallTest/orthologTest/bonobo.fa -s smallTest/orthologTest/human.fa -o output -c -t 16` appears to run smoothly.

Could it be related to my `python` environment that `wrapper.py` and `run_multiple_species_sh` are making issues?

I'm running `JustOrthologs` from a `conda` environment running `Python 2.7.15`

```
(python2) cssbbnitm026:JustOrthologs janstrauss$ pip list
DEPRECATION: Python 2.7 reached the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 is no longer maintained. pip 21.0 will drop support for Python 2.7 in January 2021. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support
Package         Version
--------------- -------------------
biopython       1.76
certifi         2019.11.28
multiprocessing 2.6.2.1
numpy           1.16.5
pip             20.1.1
setuptools      44.0.0.post20200102
wheel           0.34.2
```

Python packages `os` and `sys` that are required for `JustOrthologs` are not listed but trying to install them gives the following error:

```
(python2) cssbbnitm026:JustOrthologs janstrauss$ pip install --user os
DEPRECATION: Python 2.7 reached the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 is no longer maintained. pip 21.0 will drop support for Python 2.7 in January 2021. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support
ERROR: Could not find a version that satisfies the requirement os (from versions: none)
ERROR: No matching distribution found for os
```

Many thanks again for your help!

Jan

**Comment by ridgelab** [comment]
That's great that you were able to get JustOrthologs.py working!

The `wrapper` and `run_multiple_species.sh` were built for a Linux environment (including bash on Ubuntu on Windows). `os` and `sys` are part of the standard library, and both of them are listed in the `README`, `README_WRAPPER`, and `README_OTHER_PROGRAMS`. Since you were able to successfully run `JustOrthologs`, those libraries are already available to you. However, they do work differently in a Windows environment, which isn't currently supported. The `wrapper` facilitates parsing the gff3/fasta files. Since its job is to run the intermediate programs and make some common use cases easier, it requires being able to write and remove files. However, the programs that are called by the `wrapper` can also be called manually. In the `README_OTHER_PROGRAMS`, it lists examples of how to run each of the programs called in `wrapper` and `run_multiple_species.sh`. The `combineOrthoGroups.py` is used by run_multiple_species.sh. The other programs are used by the `wrapper`, in order: `gff3_parser.py`, `getNoException.py` (optional), `sortFastaBySeqLen.sh`, then that output is passed into `JustOrthologs`.  Your python environment, `Python 2.7.15`, should be fine.

**Comment by janstrauss1** [comment]
Hi @ridgelab,

thanks for your explanations.

I'm working in a `macOS` environment. It's a bit unfortunate that the wrapper and `run_multiple_species.sh` don't seem to run in this environment, because I was hoping to use `JustOrthologs` to detect orthologs in a large set of genome and gff annotation files downloaded from NCBI.

Still, I've tried to call each program manually as you suggest. While `gff3_parser.py` and `getNoException.py` run smoothly, I get the following error when running `sortFastaBySeqLen.sh`:

```
(python2) cssbbnitm026:JustOrthologs janstrauss$ bash sortFastaBySeqLen.sh smallTest/testNoException/test.fasta output.fasta
sed: illegal option -- r
usage: sed script [-Ealn] [-i extension] [file ...]
       sed [-Ealn] [-i extension] [-e script] ... [-f script_file] ... [file ...]
sed: illegal option -- r
usage: sed script [-Ealn] [-i extension] [file ...]
       sed [-Ealn] [-i extension] [-e script] ... [-f script_file] ... [file ...]
sed: illegal option -- r
usage: sed script [-Ealn] [-i extension] [file ...]
       sed [-Ealn] [-i extension] [-e script] ... [-f script_file] ... [file ...]
sed: illegal option -- r
usage: sed script [-Ealn] [-i extension] [file ...]
       sed [-Ealn] [-i extension] [-e script] ... [-f script_file] ... [file ...]
sed: illegal option -- r
usage: sed script [-Ealn] [-i extension] [file ...]
       sed [-Ealn] [-i extension] [-e script] ... [-f script_file] ... [file ...]
```
When I try to pass the output from `getNoException.py` to `JustOrthologs.py` instead, no ortholog groups are detected.

Any ideas how to fix this issue of `sortFastaBySeqLen.sh` to be able to get valid results?

Many thanks in advance for your help!

Jan

**Comment by ridgelab** [comment]
We just pushed an update that should fix the issue that you're experiencing. We tested on macOS without it throwing an error. The issue stems from differences between sed commands on macOS and Linux. We added another requirement to the `README_WRAPPER` as follows:
```
If you are using MacOS, make sure that gsed is installed:
    brew install gnu-sed
```
The `sortFastaBySeqLen.sh` script calls sed, but now correctly checks if the OS is macOS. If it is, then gsed is called instead. 

`sortFastaBySeqLen.sh` is crucial for the execution of JustOrthologs because JustOrthologs is optimized based on the number of CDS regions per gene and gets most of its speedup by prematurely terminating analyses based on the number of CDS regions within the sequences being analyzed (i.e., it will not compare sequences if one sequence has 2 exons and the other sequence has 20 exons, and it will not check the rest of the file for an ortholog to sequence A once it reaches its tolerance threshold because it assumes that the FASTA is sorted by the number of exons).

Please let us know if that fixes the issue for you.

**Comment by janstrauss1** [comment]
Hi @ridgelab,

many thanks - that fixes the `sed` issue thrown by `sortFastaBySeqLeng.sh`!

Also, it appears that `wrapper.py` and `run_multiple_species.sh` are now working in my macOS environment, too.

Yet, I noticed that `run_multiple_species.sh` did only do multiple pairwise comparisons of files in the input directory given by `-i` but did not combine OrthoGroups in the output, so I needed to run `combineOrthoGroups.py` subsequently. Is this expected behaviour?

Also, there seems to be a need to supply decompressed `.gff` genome annotation files, while supplying compressed `.fna.gz` genome sequence files seem to work fine. Again, is this normal behaviour or should it also be possible to supply `.gff.gz` genome annotation files to `wrapper.py` and `run_multiple_species.sh`?

Many thanks again for your help!

**Comment by ridgelab** [comment]
There was an additional sed command in `run_multiple_species.sh` that we fixed for macOS users and just pushed to the repo. That was likely causing the issue with `.gff.gz` not working for you. I just ran the tests using both compressed and decompressed .gff3 files, and they produced the same results without error (although I can't test on a Mac at the moment). You must ensure that all gff3 files are either compressed or decompressed with the same file extension (ending in .gz if it is compressed). My guess is that the sed ->gsed conversion in `run_multiple_species.sh` will fix that issue, though.
